File: JanperCab/3-Application/Dtos/DeliveryDocket/DeliveryDocketDto.cs
using _1_Domain.Enum;
using _3_Application.Dtos.Customer;
using _3_Application.Dtos.Invoice;
using System;

namespace _3_Application.Dtos.DeliveryDocket
{
    public enum DeliveryDocketType
    {
        Duraform,
        JanperEdge
    }

    public abstract class DeliveryDocketDto
    {
        public DeliveryDocketType DeliveryDocketType { get; set; }

        public int Id { get; set; }

        public int CustomerId { set; get; }

        public int? ManagerId { get; set; }

        public string CustomerReference { get; set; }

        public EnquiryTypeEnum EnquiryType { get; set; }

        public EnquiryPaymentType EnquiryPaymentType { get; set; }

        public DateTime CreatedDate { get; set; }

        public DateTime? LastEditted { get; set; }

        public DateTime? OrderedDate { get; set; }

        public DateTime? ApprovedDate { get; set; }

        public string CreatorId { get; set; }


        public string InvoiceTo { get; set; }

        public string InvoiceAddress { get; set; }

        public string InvoiceSuburb { get; set; }

        public string InvoiceState { get; set; }

        public string InvoicePostcode { get; set; }

        public string DeliveryTo { get; set; }

        public string DeliveryAddress { get; set; }

        public string DeliverySuburb { get; set; }

        public string DeliveryState { get; set; }

        public string DeliveryPostcode { get; set; }

        public string DeliveryNote { get; set; }

        public decimal GstRate { get; set; }

        public decimal DiscountRate { get; set; }

        public decimal DeliveryFee { get; set; }

        public decimal SubTotal { get; set; }

        public decimal TotalGst { get; set; }

        public decimal? TotalPrice { get; set; }

        public bool NotEditable { get; set; }

        public bool IsDeclineable { get; set; }

        public int? DeliveryRunSheetId { get; set; }


        public InvoiceDto Invoice { get; set; }

        public virtual CustomerDto Customer { get; set; }

        public virtual CustomerDto Manager { get; set; }
    }
}

File: JanperCab/5-JanperCab/ClientApp/src/app/_models/delivery-docket/DeliveryDocketDto.ts
import { Type } from 'class-transformer';
import { DeliveryDocketType } from 'src/app/_enums/DeliveryDocketType';
import { EnquiryPaymentType } from 'src/app/_enums/EnquiryPaymentType';
import { EnquiryTypeEnum } from 'src/app/_enums/EnquiryTypeEnum';
import { CabinetMakerDto } from '../customer/CabinetMakerDto';
import { CustomerDto } from '../customer/CustomerDto';
import { DistributorDto } from '../customer/DistributorDto';
import { ManufacturerDto } from '../customer/ManufacturerDto';
import { InvoiceDto } from '../invoice/InvoiceDto';

export abstract class DeliveryDocketDto {
  deliveryDocketType: DeliveryDocketType;
  id: number;
  customerReference: string;
  enquiryType: EnquiryTypeEnum;
  enquiryPaymentType: EnquiryPaymentType;
  createdDate: Date;
  lastEditted: Date;
  orderedDate: Date;
  approvedDate: Date;

  creatorId: string;

  customerId: number;
  managerId: number;

  invoiceTo: string;
  invoiceAddress: string;
  invoiceSuburb: string;
  invoiceState: string;
  invoicePostcode: string;

  deliveryTo: string;
  deliveryAddress: string;
  deliverySuburb: string;
  deliveryState: string;
  deliveryPostcode: string;

  useBlackBoard: boolean;
  deliveryNote: string;

  gstRate: number;
  discountRate: number;

  deliveryFee: number;
  subTotal: number;
  totalGst: number;
  totalPrice: number;

  notEditable: boolean;
  isDeclineable: boolean;

  deliveryRunSheetId: number;

  @Type(() => InvoiceDto)
  invoice: InvoiceDto;

  @Type(() => CustomerDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: CabinetMakerDto,
          name: '_3_Application.Dtos.Customer.CabinetMakerDto, 3-Application',
        },
        {
          value: DistributorDto,
          name: '_3_Application.Dtos.Customer.DistributorDto, 3-Application',
        },
        {
          value: ManufacturerDto,
          name: '_3_Application.Dtos.Customer.ManufacturerDto, 3-Application',
        },
      ],
    },
  })
  customer: CustomerDto;

  @Type(() => CustomerDto, {
    keepDiscriminatorProperty: true,
    discriminator: {
      property: '$type',
      subTypes: [
        {
          value: CabinetMakerDto,
          name: '_3_Application.Dtos.Customer.CabinetMakerDto, 3-Application',
        },
        {
          value: DistributorDto,
          name: '_3_Application.Dtos.Customer.DistributorDto, 3-Application',
        },
        {
          value: ManufacturerDto,
          name: '_3_Application.Dtos.Customer.ManufacturerDto, 3-Application',
        },
      ],
    },
  })
  manager: CustomerDto;
}
